File: Platforms/Windows/Reign.Video.D3D9/DepthStencil.cpp
#include "pch.h"
#include "DepthStencil.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Properties
	IDirect3DSurface9* DepthStencil::Surface::get() {return surface;}
	#pragma endregion

	#pragma region Contructors
	DepthStencil::DepthStencil(DisposableI^ parent, int width, int height, DepthStenicFormats depthStenicFormats)
	: Disposable(parent)
	{
		init(parent, width, height, depthStenicFormats);
	}

	void DepthStencil::init(DisposableI^ parent, int width, int height, DepthStenicFormats depthStenicFormats)
	{
		null();

		this->width = width;
		this->height = height;
		this->depthStenicFormats = depthStenicFormats;
		
		try
		{
			video = parent->FindParentOrSelfWithException<Video^>();

			IDirect3DSurface9* surfaceTEMP = 0;
			if (FAILED(video->Device->CreateDepthStencilSurface(width, height, D3DFMT_D24S8, D3DMULTISAMPLE_NONE, 0, false, &surfaceTEMP, 0)))
			{
				Debug::ThrowError(L"DepthStencil", L"Failed to create DepthStencilSurface");
			}
			surface = surfaceTEMP;

			if (!this->video->IsExDevice && !this->video->deviceReseting)
			{
				video->DeviceLost += gcnew Video::DeviceLostMethod(this, &DepthStencil::deviceLost);
				video->DeviceReset += gcnew Video::DeviceLostMethod(this, &DepthStencil::deviceReset);
			}
		}
		catch (Exception^ ex)
		{
			delete this;
			throw ex;
		}
	}

	DepthStencil::~DepthStencil()
	{
		disposeChilderen();
		if (video)
		{
			video->DeviceLost -= gcnew Video::DeviceLostMethod(this, &DepthStencil::deviceLost);
			video->DeviceReset -= gcnew Video::DeviceLostMethod(this, &DepthStencil::deviceReset);
		}
		dispose();
	}

	void DepthStencil::dispose()
	{
		if (surface) surface->Release();
		null();
	}

	void DepthStencil::null()
	{
		surface = 0;
	}

	void DepthStencil::deviceLost()
	{
		if (!video->IsExDevice) dispose();
	}

	void DepthStencil::deviceReset()
	{
		if (!video->IsExDevice) init(video, width, height, depthStenicFormats);
	}
	#pragma endregion

	#pragma region Methods
	void DepthStencil::enable()
	{
		video->Device->SetDepthStencilSurface(surface);
	}
	#pragma endregion
}
}
}

File: Platforms/Windows/Reign.Video.D3D9/DepthStencil.cs
using System;
using Reign.Core;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class DepthStencil : Disposable, DepthStencilI
	{
		#region Properties
		public Size2 Size {get; private set;}
		public Vector2 SizeF {get; private set;}

		internal DepthStencilCom com;
		#endregion

		#region Constructors
		public static DepthStencilI New(DisposableI parent, int width, int height, DepthStenicFormats depthStenicFormats)
		{
			return new DepthStencil(parent, width, height, depthStenicFormats);
		}

		public DepthStencil(DisposableI parent, int width, int height, DepthStenicFormats depthStenicFormats)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();
				Size = new Size2(width, height);
				SizeF = Size.ToVector2();

				com = new DepthStencilCom();
				var error = com.Init(video.com, width, height);

				switch (error)
				{
					case (DepthStencilErrors.DepthStencilSurface): Debug.ThrowError("DepthStencil", "Failed to create DepthStencil Surface"); break;
				}
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion
	}
}